[PLAT-9120] [yugabyte] Decouple istio and servicePerPod

shinji62 opened the pull request:
Decoupling service per pod and Istio
This allow to create service per pod without adding the side car injection, for example if I don't want my DB to be on my service mesh but I need to the service per pod because I would like to use XCluster without the need of pod-to-pod communication and without MCS.

Or in case of YBA been in a different namespace than YB and cross-namespace communication is not allow.

Commit:
Decouple istio and servicePerPod

shinji62 commented:
cc @iSignal

iSignal commented:
@shinji62 : servicePerPod can be technically used without Istio - what is the environment in which you are using this? 
@bhavin192 : could you review this?

bhavin192 commented on stable/yugabyte/templates/_helpers.tpl:
```diff
@@ -206,7 +206,7 @@ Get files from fs data directories for readiness / liveness probes.
 */}}
 {{- define "yugabyte.server_fqdn" -}}
-  {{- if (and .Values.istioCompatibility.enabled .Values.multicluster.createServicePerPod) -}}
+  {{- if (or .Values.istioCompatibility.enabled .Values.multicluster.createServicePerPod) -}}
```
For all these cases you can just drop the `.Values.istioCompatibility.enabled` from this condition, because all are more related to `createServicePerPod` actually.

```yaml
  {{- if .Values.multicluster.createServicePerPod -}}
```

bhavin192 reviewed (changes requested):
The change looks good, have only one small suggestion.

Commit:
Remove unecessary Istio check

shinji62 commented:
@bhavin192 Comment addressed.
@iSignal K8S based on F5 Volterra distribution.

Commit:
Remove unecessary Istio check

bhavin192 commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -386,7 +386,7 @@ spec:
           {{- $broadcastAddr := include "yugabyte.server_broadcast_address" $serviceValues -}}
           {{/* skip bind check for Istio multi-cluster, we cannot/don't bind to service IP */}}
-          {{- if (not (and $root.Values.istioCompatibility.enabled $root.Values.multicluster.createServicePerPod)) }}
+          {{- if (not (or $root.Values.istioCompatibility.enabled $root.Values.multicluster.createServicePerPod)) }}
```
We can modify the comment as we are decoupling servicePerPod and istioCompatibility. 
```
{{/* skip bind check for servicePerPod multi-cluster, we cannot/don't bind to service IP */}}
```
Also, this skip thing is specific to servicePerPod so if can be just:
```
{{- if not $root.Values.multicluster.createServicePerPod }}
```

shinji62 replied:
done

bhavin192 commented on stable/yugabyte/templates/_helpers.tpl:
```diff
@@ -267,7 +267,7 @@ Generate server web interface.
 */}}
 {{- define "yugabyte.cql_proxy_bind_address" -}}
-  {{- if or .Values.istioCompatibility.enabled .Values.multicluster.createServiceExports -}}
+  {{- if .Values.multicluster.createServiceExports -}}
```
We should not change this because when you use this chart with Istio you need the bind_address to be `0.0.0.0`. 
So, this or condition should remain as is.

shinji62 replied:
done

bhavin192 reviewed (commented):
Only two fixes, and then we can merge this PR.

Commit:
Update requested changes

shinji62 commented:
@bhavin192 Addressed your comments.

bhavin192 commented:
@shinji62 have you tested these changes on a cluster? Just tried doing `helm template --set multicluster.createServicePerPod=true .` and it seems to be setting the rpc_bind_address to the service FQDN which won't work. 
Basically, just setting createServicePerPod with this change might not work. We might need to modify the rpc_bind_address to be $(POD_IP) I think? Need to test that. 
This template will need a change
```
{{- define "yugabyte.rpc_bind_address" -}}
```

Commit:
Chnage RPC address

shinji62 commented:
@bhavin192 I did test before the first time I submit the PR, not every change, honestly I don't know much about what is expected for the rpc bind .. but I did the same as itio, pod ip or `127.0.0.1` etc ..

bhavin192 commented on stable/yugabyte/templates/_helpers.tpl:
```diff
@@ -243,7 +243,7 @@ we stick to 0.0.0.0, which works for master.
 {{- define "yugabyte.rpc_bind_address" -}}
   {{- $port := index .Service.ports "tcp-rpc-port" -}}
-  {{- if .Values.istioCompatibility.enabled -}}
+  {{- if (or .Values.istioCompatibility.enabled .Values.multicluster.createServicePerPod) -}}
```
It would be better to set it to `$(POD_IP):{{ $port }}` i.e. same as createServiceExports. 
That seems to be a overall better option than 0.0.0.0.

shinji62 replied:
done @bhavin192

Commit:
Change to pod ip port

baba230896 committed:
Fixed the cql_proxy_bind_address.

Test
1.
multicluster:
  createServicePerPod: true

tls:
  enabled: true

isMultiAz: true
masterAddresses: "yb-master-0.yb-platform.svc.cluster.local,yb-master-1.yb-platform.svc.cluster.local,yb-master-2.yb-platform.svc.cluster.local"

2.
multicluster:
  createServicePerPod: true

bhavin192 reviewed (approved):
Yatish has tested these changes. Only one thing to note here is when using the YB chart without YBA one has to set isMultiAZ=true, and set the masterAddresses to correct value to get this working. In case of YBA these are taken care of.
